cloudinit/sources/__init__.py

```diff
@@ -307,6 +307,29 @@ def _unpickle(self, ci_pkl_version: int) -> None:
     def __str__(self):
         return type_utils.obj_name(self)
 
+    def ds_detect(self) -> bool:
+        """Check if running on this datasource"""
+        return True
+
+    def override_ds_detect(self):
+        """Override if either:
+        - only a single datasource defined (nothing to fall back to)
+        - TODO: commandline argument is used (ci.ds=OpenStack)
+        """
+        return self.sys_cfg.get("datasource_list", []) in (
+            [self.dsname],
+            [self.dsname, "None"],
+        )
+
+    def _check_and_get_data(self):
+        """Overrides runtime datasource detection"""
+        if self.override_ds_detect() or self.ds_detect():
+            LOG.debug("Machine is running on %s.", self)
```

Comment:
??? WDYT
```suggestion
            LOG.debug("Machine configured to run on single datasource %s.", self)
```

Reply:
As written, this is incorrect. I think you mean for that log to be written only if `self.override_ds_detect()`, but not if `self.ds_detect()`?

I can rework the logic to make that happen; I can see that being a useful distinction to log.